mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitorV2.java

```diff
@@ -0,0 +1,809 @@
+
+/** Immutable class representing the state of a device's network. */
+class NetworkState {
+  private final boolean mConnected;
+  private final int mType;
+  private final int mSubtype;
+  private final boolean mIsMetered;
+  // WIFI SSID of the connection on pre-Marshmallow, NetID starting with Marshmallow. Always
+  // non-null (i.e. instead of null it'll be an empty string) to facilitate .equals().
+  private final String mNetworkIdentifier;
+  // Indicates if this network is using DNS-over-TLS.
+  private final boolean mIsPrivateDnsActive;
+  // Indicates the DNS-over-TLS server in use, if specified.
+  private final String mPrivateDnsServerName;
+
+  // Consolidate network type and subtype into one enum.
+  public static EnvoyConnectionType convertToEnvoyConnectionType(int type, int subtype) {
+    switch (type) {
+    case ConnectivityManager.TYPE_ETHERNET:
+      return EnvoyConnectionType.CONNECTION_ETHERNET;
+    case ConnectivityManager.TYPE_WIFI:
+      return EnvoyConnectionType.CONNECTION_WIFI;
+    case ConnectivityManager.TYPE_WIMAX:
+      return EnvoyConnectionType.CONNECTION_4G;
+    case ConnectivityManager.TYPE_BLUETOOTH:
+      return EnvoyConnectionType.CONNECTION_BLUETOOTH;
+    case ConnectivityManager.TYPE_MOBILE:
+    case ConnectivityManager.TYPE_MOBILE_DUN:
+    case ConnectivityManager.TYPE_MOBILE_HIPRI:
+      // Use information from TelephonyManager to classify the connection.
+      switch (subtype) {
+      case TelephonyManager.NETWORK_TYPE_GPRS:
+      case TelephonyManager.NETWORK_TYPE_EDGE:
+      case TelephonyManager.NETWORK_TYPE_CDMA:
+      case TelephonyManager.NETWORK_TYPE_1xRTT:
+      case TelephonyManager.NETWORK_TYPE_IDEN:
+        return EnvoyConnectionType.CONNECTION_2G;
+      case TelephonyManager.NETWORK_TYPE_UMTS:
+      case TelephonyManager.NETWORK_TYPE_EVDO_0:
+      case TelephonyManager.NETWORK_TYPE_EVDO_A:
+      case TelephonyManager.NETWORK_TYPE_HSDPA:
+      case TelephonyManager.NETWORK_TYPE_HSUPA:
+      case TelephonyManager.NETWORK_TYPE_HSPA:
+      case TelephonyManager.NETWORK_TYPE_EVDO_B:
+      case TelephonyManager.NETWORK_TYPE_EHRPD:
+      case TelephonyManager.NETWORK_TYPE_HSPAP:
+        return EnvoyConnectionType.CONNECTION_3G;
+      case TelephonyManager.NETWORK_TYPE_LTE:
+        return EnvoyConnectionType.CONNECTION_4G;
+      case TelephonyManager.NETWORK_TYPE_NR:
+        return EnvoyConnectionType.CONNECTION_5G;
+      default:
+        return EnvoyConnectionType.CONNECTION_UNKNOWN;
+      }
+    default:
+      return EnvoyConnectionType.CONNECTION_UNKNOWN;
+    }
+  }
+
+  public NetworkState(boolean connected, int type, int subtype, boolean isMetered,
+                      String networkIdentifier, boolean isPrivateDnsActive,
+                      String privateDnsServerName) {
+    mConnected = connected;
+    mType = type;
+    mSubtype = subtype;
+    mIsMetered = isMetered;
+    mNetworkIdentifier = networkIdentifier == null ? "" : networkIdentifier;
```

Comment:
I'm always confused about using null v.s. Optional in Java, why prefer Optional over null here?

Reply:
In Java, I think the preferred way to express whether a value may or may not be present is through `Optional`.  It also forces using the methods, so you have to get the value via `get()` or `orElse()` with a default value.  

That said, it seems like `NetworkState` is only used in this class, so it probably doesn't add much value to make the class take in an Optional as a parameter.  However, since NetworkState isn't used outside of this class, we should make the class `private`.